### ISSUE TITLE
crash when selecting large images

### ISSUE THREAD

**Issue by maxnemoy** [Sh1d0w/multi_image_picker#44]
if i select images with size> 7mb the application is terminated

log 
```
Installing build\app\outputs\apk\app.apk...
Syncing files to device Custom Phone...
D/        ( 4773): HostConnection::get() New Host Connection established 0xe3e1d9c0, tid 4796
W/art     ( 4773): Before Android 4.1, method android.graphics.PorterDuffColorFilter android.support.graphics.drawable.VectorDrawableCompat.updateTintFilter(android.graphics.PorterDuffColorFilter, android.content.res.ColorStateList, android.graphics.PorterDuff$Mode) would have incorrectly overridden the package-private method in android.graphics.drawable.Drawable
E/EGL_emulation( 4773): tid 4805: eglSurfaceAttrib(1174): error 0x3009 (EGL_BAD_MATCH)
W/OpenGLRenderer( 4773): Failed to set EGL_SWAP_BEHAVIOR on surface 0xc82fff80, error=EGL_BAD_MATCH
I/ViewConfigCompat( 4773): Could not find method getScaledScrollFactor() on ViewConfiguration
E/EGL_emulation( 4773): tid 4805: eglSurfaceAttrib(1174): error 0x3009 (EGL_BAD_MATCH)
W/OpenGLRenderer( 4773): Failed to set EGL_SWAP_BEHAVIOR on surface 0xc9575300, error=EGL_BAD_MATCH
D/OpenGLRenderer( 4773): endAllActiveAnimators on 0xcd1ec900 (AppCompatTextView) with handle 0xc66249f0
I/art     ( 4773): Starting a blocking GC Alloc
I/art     ( 4773): Starting a blocking GC Alloc
I/art     ( 4773): Alloc sticky concurrent mark sweep GC freed 2654(159KB) AllocSpace objects, 4(80KB) LOS objects, 37% free, 2MB/4MB, paused 660us total 15.317ms
I/art     ( 4773): Starting a blocking GC Alloc
I/art     ( 4773): Alloc partial concurrent mark sweep GC freed 2227(97KB) AllocSpace objects, 0(0B) LOS objects, 40% free, 2MB/4MB, paused 331us total 46.677ms
I/art     ( 4773): Starting a blocking GC Alloc
I/art     ( 4773): Alloc concurrent mark sweep GC freed 106(4KB) AllocSpace objects, 0(0B) LOS objects, 39% free, 2MB/4MB, paused 483us total 25.602ms
I/art     ( 4773): Forcing collection of SoftReferences for 195MB allocation
I/art     ( 4773): Starting a blocking GC Alloc
I/art     ( 4773): Alloc concurrent mark sweep GC freed 4(88B) AllocSpace objects, 0(0B) LOS objects, 40% free, 2MB/4MB, paused 1.299ms total 26.081ms
W/art     ( 4773): Throwing OutOfMemoryError "Failed to allocate a 205514212 byte allocation with 1777584 free bytes and 93MB until OOM"
I/art     ( 4773): Starting a blocking GC Alloc
I/art     ( 4773): Starting a blocking GC Alloc
I/art     ( 4773): Alloc sticky concurrent mark sweep GC freed 4(464B) AllocSpace objects, 0(0B) LOS objects, 39% free, 2MB/4MB, paused 247us total 7.639ms
I/art     ( 4773): Starting a blocking GC Alloc
I/art     ( 4773): Alloc partial concurrent mark sweep GC freed 8(176B) AllocSpace objects, 0(0B) LOS objects, 39% free, 2MB/4MB, paused 200us total 9.527ms
I/art     ( 4773): Starting a blocking GC Alloc
I/art     ( 4773): Alloc concurrent mark sweep GC freed 4(88B) AllocSpace objects, 0(0B) LOS objects, 40% free, 2MB/4MB, paused 276us total 36.779ms
I/art     ( 4773): Forcing collection of SoftReferences for 195MB allocation
I/art     ( 4773): Starting a blocking GC Alloc
I/art     ( 4773): Alloc concurrent mark sweep GC freed 5(104B) AllocSpace objects, 0(0B) LOS objects, 39% free, 2MB/4MB, paused 170us total 17.814ms
W/art     ( 4773): Throwing OutOfMemoryError "Failed to allocate a 205514212 byte allocation with 1777572 free bytes and 93MB until OOM"
E/AndroidRuntime( 4773): FATAL EXCEPTION: AsyncTask #2
E/AndroidRuntime( 4773): Process: com.example.hungerproject, PID: 4773
E/AndroidRuntime( 4773): java.lang.RuntimeException: An error occurred while executing doInBackground()
E/AndroidRuntime( 4773): 	at android.os.AsyncTask$3.done(AsyncTask.java:325)
E/AndroidRuntime( 4773): 	at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:354)
E/AndroidRuntime( 4773): 	at java.util.concurrent.FutureTask.setException(FutureTask.java:223)
E/AndroidRuntime( 4773): 	at java.util.concurrent.FutureTask.run(FutureTask.java:242)
E/AndroidRuntime( 4773): 	at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:243)
E/AndroidRuntime( 4773): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
E/AndroidRuntime( 4773): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
E/AndroidRuntime( 4773): 	at java.lang.Thread.run(Thread.java:761)
E/AndroidRuntime( 4773): Caused by: java.lang.OutOfMemoryError: Failed to allocate a 205514212 byte allocation with 1777572 free bytes and 93MB until OOM
E/AndroidRuntime( 4773): 	at dalvik.system.VMRuntime.newNonMovableArray(Native Method)
E/AndroidRuntime( 4773): 	at android.graphics.BitmapFactory.nativeDecodeStream(Native Method)
E/AndroidRuntime( 4773): 	at android.graphics.BitmapFactory.decodeStreamInternal(BitmapFactory.java:646)
E/AndroidRuntime( 4773): 	at android.graphics.BitmapFactory.decodeStream(BitmapFactory.java:622)
E/AndroidRuntime( 4773): 	at android.graphics.BitmapFactory.decodeStream(BitmapFactory.java:660)
E/AndroidRuntime( 4773): 	at com.vitanov.multiimagepicker.MultiImagePickerPlugin.getCorrectlyOrientedImage(MultiImagePickerPlugin.java:595)
E/AndroidRuntime( 4773): 	at com.vitanov.multiimagepicker.MultiImagePickerPlugin.access$300(MultiImagePickerPlugin.java:61)
E/AndroidRuntime( 4773): 	at com.vitanov.multiimagepicker.MultiImagePickerPlugin$GetThumbnailTask.doInBackground(MultiImagePickerPlugin.java:151)
E/AndroidRuntime( 4773): 	at com.vitanov.multiimagepicker.MultiImagePickerPlugin$GetThumbnailTask.doInBackground(MultiImagePickerPlugin.java:128)
E/AndroidRuntime( 4773): 	at android.os.AsyncTask$2.call(AsyncTask.java:305)
E/AndroidRuntime( 4773): 	at java.util.concurrent.FutureTask.run(FutureTask.java:237)
E/AndroidRuntime( 4773): 	... 4 more
Lost connection to device.
```

flutter doctor

```
[√] Flutter (Channel master, v1.1.2-pre.36, on Microsoft Windows [Version 10.0.17763.195], locale ru-RU)
[√] Android toolchain - develop for Android devices (Android SDK 28.0.3)
[√] Android Studio (version 3.2)
[√] IntelliJ IDEA Ultimate Edition (version 2018.3)
[√] Connected device (1 available)

• No issues found!


```

**Comment by Sh1d0w** [comment]
Please provide example dart code you are using.

If you are requesting a lot of original images at the same time to display them without properly managing the memory this can occur.

In general I would suggest you if you want to display the images to use `requestThumbnail(width, height, quality)` method with a quality parameter as well (this will reduce the memory allocated from 7mb to several KB). When thumbs are not needed anymore call `release()` method on the Asset to release the allocated memory.

Use `requestOriginal` only when you need the actual original image and then again when it is not needed, release it.

**Comment by maxnemoy** [comment]
```
void _loadImage() async {
    await this._asset.requestThumbnail(200, 200, quality: 50);
    setState(() {});
  }
```

**Comment by maxnemoy** [comment]
I am using requestThumbnail

when requesting images size =<5mb
no problem.

Can I limit the choice of large images for a user?

**Comment by Sh1d0w** [comment]
@maxnemoy Are you testing this with the app in the example folder, or you have written your own? I am asking because I am testing it with the example code, and I can successfully pick 20 images, which are all 15 MB in size?

If you can provide the full source code in a gist so I can reproduce this?

**Comment by maxnemoy** [comment]
```
import 'dart:async';
import 'dart:io';
import 'dart:math' as Math;

import 'package:flutter/cupertino.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:image/image.dart' as Im;
import 'package:multi_image_picker/asset.dart';
import 'package:multi_image_picker/multi_image_picker.dart';
import 'package:path_provider/path_provider.dart';
import 'package:flutter_native_image/flutter_native_image.dart';

class TestPage extends StatefulWidget {
  final Function createProduct;
  final String currentProject;
  final String currentCategory;

  TestPage({this.createProduct, this.currentCategory, this.currentProject});

  @override
  _TestPageState createState() => _TestPageState();
}

class _TestPageState extends State<TestPage> {
  List<Asset> images = List<Asset>();
  String _error;

  Future<void> loadAssets() async {
    setState(() {
      images = List<Asset>();
    });

    List<Asset> resultList;
    String error;

    try {
      resultList = await MultiImagePicker.pickImages(
        maxImages: 300,
        enableCamera: true,
        //options: CupertinoOptions(takePhotoIcon: "chat"),
      );
    } on PlatformException catch (e) {
      error = e.message;
    }

    // If the widget was removed from the tree while the asynchronous platform
    // message was in flight, we want to discard the reply rather than calling
    // setState to update our non-existent appearance.
    if (!mounted) return;

    setState(() {
      images = resultList;
      if (error == null) _error = 'No Error Dectected';
    });
  }

  Widget buildGridView() {
    return Container(
      height: 200.0,
      child: ListView.builder(
          scrollDirection: Axis.horizontal,
          itemCount: images.length,
          itemBuilder: (BuildContext context, index) =>
              ImageItemList(index, images[index])),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Column(
        children: <Widget>[
          RaisedButton(
            child: Text("Pick images"),
            onPressed: loadAssets,
          ),
          buildGridView(),
        ],
      ),
    );
  }
}

class ImageItemList extends StatefulWidget {
  final int _index;
  final Asset _asset;

  ImageItemList(this._index, this._asset);

  @override
  State<StatefulWidget> createState() =>
      _ImageItemListState(this._index, this._asset);
}

class _ImageItemListState extends State<ImageItemList> {
  int _index = 0;
  Asset _asset;

  _ImageItemListState(this._index, this._asset);

  @override
  void initState() {
    super.initState();
    _loadImage();
  }

  static void savePic(Pic picture){

    File(picture.path)..writeAsBytesSync(Im.encodeJpg(Im.decodeImage(picture.img), quality: 50));
  }




  void _loadImage() async {
      await this._asset.requestThumbnail(10, 10, quality: 50);


//    await this._asset.requestOriginal(quality: 50);
/////////////////////////////////
//    final tempDir = await getApplicationDocumentsDirectory();
//    String path = tempDir.path;
//
//    int rand = new Math.Random().nextInt(10000);
//    String pathImage = '$path/img_$rand.jpg';
//    await _asset.requestOriginal(quality: 100);
//    List<int> img =_asset.imageData.buffer.asUint8List();
//    compute(savePic, Pic(img: img, path: pathImage));

///////////////////////////

    setState(() {});
  }

  @override
  Widget build(BuildContext context) {
    if (null != this._asset.thumbData) {
      return Padding(
        padding: const EdgeInsets.all(8.0),
        child: GestureDetector(
          onTap: () {
            print("open index $_index");
          },
          child: Container(
            width: 200,
            // color: Colors.blueGrey,
            child: Stack(
              children: <Widget>[
                Align(
                    alignment: Alignment.center,
                    child: Image.memory(
                      this._asset.thumbData.buffer.asUint8List(),
                      fit: BoxFit.cover,
                      gaplessPlayback: true,
                    )),
                Align(
                    alignment: Alignment.topRight,
                    child: IconButton(
                      icon: Icon(Icons.remove_circle_outline,
                          color: Colors.red[800]),
                      onPressed: () {
                        print("remove" + _index.toString());
                      },
                    ))
              ],
            ),
          ),
        ),
      );
    }

    return Padding(
      padding: const EdgeInsets.all(8.0),
      child: GestureDetector(
        onTap: () {
          print("open index $_index");
        },
        child: Container(
          width: 200,
          // color: Colors.blueGrey,
          child: Stack(
            children: <Widget>[
              Align(
                  alignment: Alignment.center,
                  child: CircularProgressIndicator()),
              Align(
                alignment: Alignment.bottomCenter,
                child: Padding(
                  padding: const EdgeInsets.all(5.0),
                  child: Text("image loading.."),
                ),
              )
            ],
          ),
        ),
      ),
    );
  }
}

class Pic{
  List<int> img;
  String path;
  Pic({this.img, this.path});
}

```

emul:
android 9
cpu 2
1gb

**Comment by Sh1d0w** [comment]
Ok, I think I know where the memory leak occurs, will let you know once I release a fix for this.

Thanks for reporting the issue.

**Comment by Sh1d0w** [comment]
Hi, I've released version 2.3.27 which fixes several memory leaks. Can you try and see if this fixes the problem for you?

I've tested with this raw image 

https://mars.nasa.gov/images/Curiosity-Rover-Portrait-Mars-Mojave-Selfie-pia19142-MALHI-full.jpg

Which is 34 MB and  18 029 x 9233 in resolution, and did not have a crash.

Also I see in your example you tried for request the original image and save it to a file - this is ok but its place is not in this widget, as this will be called for every image you pick. When you request the original  image you want to do it in a sequence where you control the amount of `requestOriginal` calls, as if they are called simultaneously you will run out of memory very quickly.

Also after you save the file, it is good to call `release()` method on the asset if you no longer need the image data, again to free up some memory.

**Comment by maxnemoy** [comment]
Thanks version 2.3.27 fix problem.

ps: i have not much experience in mobile development. I will try to write more optimal code
